### ISSUE TITLE
Application using eye-to-hand calibrated arm.

### ISSUE THREAD

**Issue by jarvishou829** [f3rm/f3rm#15]
Thanks for your great work, and we met some problems when applying. We calibrate the arm using eye-to-hand way. Also, we record the tabletop scene including a frame from the fixed camera for calibration. We want to compute the NeRF-to-world json file now. As far as we know, the NeRF-to-world means the transformation from NeRF to robot arm base, and the calibration shows the transformation from fixed camera to base. So we need to compute the transformation from the NeRF to the fixed camera now. And it is provided by running colmap, and showed in the transforms.json and data_paser.json. So we first need to recover the indeed pose of the fixed camera using data_paser.json and compute the transformation from the NeRF to the fixed camera using transforms.json. We wonder if this is the correct way.

**Comment by williamshen-nz** [comment]
Hi, we recently replicated F3RM on a new robot with a wrist camera. Here are the steps we took to calibrate the camera and get it working with pose optimization.

1. Calibrate the wrist camera (eye-in-hand calibration), this gives `cam_to_world`.
2. Scan the scene using the robot, and then run COLMAP to get the `transforms.json`. For the same trajectory, we create a `transforms_gt.json` using the calibration from `cam_to_world`.
3. Determine the scale, rotation and translation to get the `colmap_to_world` transform.
4. Train a NeRF/feature field of the scene. Nerfstudio applies it's own scaling, rotation and translation. This is in `dataparser_transforms.json` which we can use to construct the `nerf_to_colmap` transform. 
   - When debugging, you can turn off the scaling, rotation and translation and then slowly turn each on one when you get what you expect.
5. Use the `nerf_to_colmap` transform with the pose optimization code.

So in total there are 3 coordinate systems you need to deal with:
1. World
2. COLMAP
3. NeRF

Hope that helps! Let me know if you have any questions.

**Comment by dajiebi** [comment]
About Step 3，here is my initial understanding of your approach :
1.  Get colmap_frame and gruth_true_frame from transforms.json and transforms_gt.json
2.  scaling = dist (real_iamge1, real_image2) /dist (colmap_iamge1, colmap_image2)
3.  After getting scaling, scale the colmap frame to get scaled_colmap_frame
4.  Calculate the colmap_to_world with scaled_colmap_frame and gruth_true_frame
The results don't seem to be very good, where are my mistakes？

**Comment by williamshen-nz** [comment]
Here is the hacky script we used for computing `colmap_to_world` from COLMAP `transforms.json` and real `transforms_gt.json`. We do not provide support for this script. Hope it is helpful: https://gist.github.com/williamshen-nz/da87ee47a5fc3b5ebc3e0f7f777d7477

Assuming COLMAP has successfully recovered poses and your original eye-to-hand calibration was accurate, you should get translational errors < 0.01m and rotational errors < 2 degrees.

**Comment by dajiebi** [comment]
Thanks for your careful guidance, I got colmap_to_world.json now！ but I still met some problems. I don't know how to calculate nerf_to_world.json from colmap_to_world.json and dataparsere_transform.json. I made a few attempts, but it seems to fail.

**Comment by williamshen-nz** [comment]
Our original calibration pipeline didn't use dataparser_transforms.json as we calibrated end-to-end. 

Here is one way of getting things working by disabling the auto-orientation and centering during training while keeping auto pose scaling on. That way the `dataparser_transforms.json` contains only scaling. Note you should enable centering, so everything fits better in the unit cube. The code below does not do that, but it should be easy to figure out the inverse transformation.

1. Train with these additional flags `--orientation-method none --auto-scale-poses True --center-method none`. Example full command:

```bash
ns-train f3rm --max-num-iterations $TRAIN_STEPS --output_dir f3rm_outputs --timestamp f3rm \
      --viewer.quit-on-train-completion True --experiment-name $DIR_BASENAME \
      nerfstudio-data --data $SCAN_DIR \
      --orientation-method none --auto-scale-poses True --center-method none
```

The matrix in `dataparser_transforms.json` should then be the identity.

3. Modify the [load.py](https://github.com/f3rm/f3rm/blob/main/f3rm_robot/load.py#L103-L105) file where `nerf_to_world` is loaded. Here's some example code you will likely need to modify:

```python
def load_colmap_to_world(dataset: Path) -> Transform3d:
    col2w_path = dataset / "colmap_to_world.json"
    if not col2w_path.exists():
        raise ValueError(f"Could not find colmap_to_world.json in {dataset}")

    with col2w_path.open("r") as f:
        colmap_to_world_dict = json.load(f)

    colmap_to_world = colmap_to_world_dict["transformation_matrix"]
    colmap_to_world = torch.tensor(colmap_to_world).float()
    assert colmap_to_world.shape == (4, 4), f"Expected colmap_to_world to be 4x4, but got {colmap_to_world.shape}"

    colmap_to_world = Transform3d(matrix=colmap_to_world.T)
    return colmap_to_world


# Load colmap_to_world
colmap_to_world = load_colmap_to_world(dataset=config.data or config.pipeline.datamanager.dataparser.data)

# Load dataparser_transforms
dataparser_transforms = Path(exp_config_path).parent / "dataparser_transforms.json"
if not dataparser_transforms.exists():
    raise FileNotFoundError

with dataparser_transforms.open("r") as f:
    dataparser_transforms = json.load(f)

scale = dataparser_transforms["scale"]
inverse_scale = 1 / scale

scale_matrix = np.eye(4)
scale_matrix[0, 0] = inverse_scale
scale_matrix[1, 1] = inverse_scale
scale_matrix[2, 2] = inverse_scale

# Assumes poses are scaled only
# Not implemented: rotation and translation
inverse_transform = np.eye(4)
nerf_to_colmap = inverse_scale

nerf_to_colmap = torch.tensor(nerf_to_colmap).float()
print("nerf_to_colmap =", nerf_to_colmap)

# Need to take transpose as Transform3d uses row vector convention rather than column
nerf_to_colmap = Transform3d(matrix=nerf_to_colmap.T)
nerf_to_world = nerf_to_colmap.compose(colmap_to_world)
```

If everything works correctly you should see the sampled point cloud in the viewer in the expected world coordinate frame.

**Comment by xmulee** [comment]
Hello, I am currently facing a problem. I used the multi-view reconstruction method (Dust3R) to obtain point cloud data, but the point cloud data is off by a scale relative to the world coordinate system. The scale information is important for my ultimate goal of using the point cloud for measurement. Is there a way to recover the scale using the method you mentioned, or are there any other solutions?

**Comment by williamshen-nz** [comment]
If you know the camera poses in the world coordinate system, you can compute the scale. Similarly, if you have RGB-D you might be able to compute the scale that best matches the depth image (which you could take from Dust3R or backproject). You will need some way to reference what comes out of Dust3R back to the world frame.

**Comment by xmulee** [comment]
Thank you for your response. Are the camera poses obtained from COLMAP in the world coordinate system? Can the scale be calculated using the camera poses obtained from Dust3R and COLMAP?

**Comment by dajiebi** [comment]
maybe this is helpful: https://gist.github.com/williamshen-nz/da87ee47a5fc3b5ebc3e0f7f777d7477




> Here is the hacky script we used for computing `colmap_to_world` from COLMAP `transforms.json` and real `transforms_gt.json`. We do not provide support for this script. Hope it is helpful: https://gist.github.com/williamshen-nz/da87ee47a5fc3b5ebc3e0f7f777d7477
> 
> Assuming COLMAP has successfully recovered poses and your original eye-to-hand calibration was accurate, you should get translational errors < 0.01m and rotational errors < 2 degrees.

**Comment by xmulee** [comment]
How can I obtain the actual transforms_gt.json file? Thank you for your response.

**Comment by dajiebi** [comment]
I put a camera on the end of the manipulator and recorded the camera pose when each photo was taken. You should calibrate to get the transform between the camera and the manipulator flange, then get the flange pose from the manipulator when taking the photo, and then convert it to the camera pose(in world coordinate system).

**Comment by xmulee** [comment]
> If you know the camera poses in the world coordinate system, you can compute the scale. Similarly, if you have RGB-D you might be able to compute the scale that best matches the depth image (which you could take from Dust3R or backproject). You will need some way to reference what comes out of Dust3R back to the world frame.

Thank you for your response.How can I recover the scale from depth information if I already have the depth map of an image?

**Comment by dajiebi** [comment]
This seems to have been discussed here, hope this is helpful to you




> If you want metric scaled exports, you need metric scaled poses. To do this, you need to use RGBD SLAM or some other technique that uses **depth** data when estimating camera poses. If you have a recent iPhone with a depth sensor (time of flight sensor), you can get metric poses using e.g. https://www.spectacularai.com/mapping or other apps like PolyCam

**Comment by dajiebi** [comment]
https://github.com/nerfstudio-project/nerfstudio/issues/2924#issuecomment-1950225312

**Comment by xmulee** [comment]
Thank you for your reply, I understand the problem now. Thanks again.

**Comment by dajiebi** [comment]
> 我们最初的校准管道在进行端到端校准时没有使用 dataparser_transforms.json。
> 
> 这是通过在训练期间禁用自动方向和居中同时保持自动姿势缩放打开的一种方法。这样 将仅包含缩放。请注意，您应该启用居中，以便所有内容都更适合单位立方体。下面的代码没有这样做，但应该很容易找出逆变换。`dataparser_transforms.json`
> 
> 1. 使用这些附加标志进行训练。full 命令示例：`--orientation-method none --auto-scale-poses True --center-method none`
> 
> ```shell
> ns-train f3rm --max-num-iterations $TRAIN_STEPS --output_dir f3rm_outputs --timestamp f3rm \
>       --viewer.quit-on-train-completion True --experiment-name $DIR_BASENAME \
>       nerfstudio-data --data $SCAN_DIR \
>       --orientation-method none --auto-scale-poses True --center-method none
> ```
> 
> 然后，in 中的矩阵应该是恒等式。`dataparser_transforms.json`
> 
> 3. 修改 [load.py](https://github.com/f3rm/f3rm/blob/main/f3rm_robot/load.py#L103-L105) 文件 加载的位置。以下是您可能需要修改的一些示例代码：`nerf_to_world`
> 
> ```python
> def load_colmap_to_world(dataset: Path) -> Transform3d:
>     col2w_path = dataset / "colmap_to_world.json"
>     if not col2w_path.exists():
>         raise ValueError(f"Could not find colmap_to_world.json in {dataset}")
> 
>     with col2w_path.open("r") as f:
>         colmap_to_world_dict = json.load(f)
> 
>     colmap_to_world = colmap_to_world_dict["transformation_matrix"]
>     colmap_to_world = torch.tensor(colmap_to_world).float()
>     assert colmap_to_world.shape == (4, 4), f"Expected colmap_to_world to be 4x4, but got {colmap_to_world.shape}"
> 
>     colmap_to_world = Transform3d(matrix=colmap_to_world.T)
>     return colmap_to_world
> 
> 
> # Load colmap_to_world
> colmap_to_world = load_colmap_to_world(dataset=config.data or config.pipeline.datamanager.dataparser.data)
> 
> # Load dataparser_transforms
> dataparser_transforms = Path(exp_config_path).parent / "dataparser_transforms.json"
> if not dataparser_transforms.exists():
>     raise FileNotFoundError
> 
> with dataparser_transforms.open("r") as f:
>     dataparser_transforms = json.load(f)
> 
> scale = dataparser_transforms["scale"]
> inverse_scale = 1 / scale
> 
> scale_matrix = np.eye(4)
> scale_matrix[0, 0] = inverse_scale
> scale_matrix[1, 1] = inverse_scale
> scale_matrix[2, 2] = inverse_scale
> 
> # Assumes poses are scaled only
> # Not implemented: rotation and translation
> inverse_transform = np.eye(4)
> nerf_to_colmap = inverse_scale
> 
> nerf_to_colmap = torch.tensor(nerf_to_colmap).float()
> print("nerf_to_colmap =", nerf_to_colmap)
> 
> # Need to take transpose as Transform3d uses row vector convention rather than column
> nerf_to_colmap = Transform3d(matrix=nerf_to_colmap.T)
> nerf_to_world = nerf_to_colmap.compose(colmap_to_world)
> ```
> 
> 如果一切正常，您应该会在查看器中预期的世界坐标系中看到采样的点云。

I apologize for bothering you again. I really want to know how to get the correct nerf_to_world without disabling scaling, rotation, and translation.

**Comment by williamshen-nz** [comment]
Hi, you will need to reverse the transformations applied here by the nerfstudio dataparser: https://github.com/nerfstudio-project/nerfstudio/blob/v0.3.4/nerfstudio/data/dataparsers/nerfstudio_dataparser.py#L233-L246

Scale is already handled by the code, so I suggest you try enable center method next so translation is enabled. Once you get that working, you can figure out rotation. It is definitely possible, just might require some fiddling around with the order of transformations.
